### PR TITLE
feat: build hash from file sources in config

### PR DESCRIPTION
**Pull request by StarpTech**
#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: packages/sdk/src/definition/index.ts [hunk 1]
```diff
@@ -241,6 +241,11 @@ interface GraphQLIntrospectionOptions {
 	skipRenameRootFields?: string[];
 }
 
+// TODO: better name for "internal" ?
```
Comment (StarpTech) [inline packages/sdk/src/definition/index.ts:244]: Hint

File: packages/sdk/src/definition/introspection-cache.ts [hunk 2]
```diff
@@ -124,12 +125,18 @@ export const introspectInInterval = async <Introspection extends IntrospectionCo
...
-export const introspectWithCache = async <Introspection extends IntrospectionConfiguration, A extends ApiType>(
+export const introspectWithCache = async <Introspection extends IntrospectionUnion, A extends ApiType>(
 	introspection: Introspection,
 	generator: (introspection: Introspection) => Promise<Api<A>>
 ): Promise<Api<A>> => {
 	const cacheKey = objectHash(introspection);
 
+	if (introspection.kind == 'openapi') {
+		if (introspection.source.kind === 'file') {
+			// TODO: read file, hash file and combine it with cacheKey
```
Comment (StarpTech) [inline packages/sdk/src/definition/introspection-cache.ts:136]: @fiam, here you can handle all properties pointing to external files and respect them in the final hash.

**Comment by fiam** [comment]
@StarpTech Please, don't mix formatting changes with actual code changes, it makes telling what's actually changed more complicated.

If we wanted to go with this approach, I would instead make `introspectWithCache()` take an additional optional argument that indicates which files to consider for hashing, then make `introspect.openapi()` pass those. That way we remove all the plumbing for wrapping other introspection types.

However, as we discussed this morning, this simplest approach might be to skip the cache altogether for openapi data sources. What do you think of something like this?

```diff
diff --git a/packages/sdk/src/definition/index.ts b/packages/sdk/src/definition/index.ts
index 83232bc..9df9603 100644
--- a/packages/sdk/src/definition/index.ts
+++ b/packages/sdk/src/definition/index.ts
@@ -603,11 +603,10 @@ export const introspect = {
                        return new MongoDBApi(schema, dataSources, fields, types, interpolateVariableDefinitionAsJSON);
                }),
        federation: introspectFederation,
-       openApi: async (introspection: OpenAPIIntrospection): Promise<RESTApi> =>
-               introspectWithCache(introspection, async (introspection: OpenAPIIntrospection): Promise<RESTApi> => {
-                       const spec = loadOpenApi(introspection);
-                       return await openApiSpecificationToRESTApiObject(spec, introspection);
-               }),
+       openApi: async (introspection: OpenAPIIntrospection): Promise<RESTApi> => {
+               const spec = loadOpenApi(introspection);
+               return await openApiSpecificationToRESTApiObject(spec, introspection);
+       },
 };
 ```

We lose support for polling by doing that, but maybe the correct way handle schemas from static files is to make `wunderctl` watch them. What about watching `.wundergraph` and excluding `generated`/ `cache`? That would also remove a lot of complexity from the watcher.

**Comment by StarpTech** [comment]
I didn't format the code without reasons. It's reformatted based on my code changes.

We should not remove the polling functionality for openapi. The easiest approach for now is to skip the cache for openapi with file as a source and regernerate the config.

**Comment by fiam** [comment]
> I didn't format the code without reasons. It's reformatted based on my code changes.
> 
> We should not remove the polling functionality for openapi. The easiest approach for now is to skip the cache for openapi with file as a source and regernerate the config.

You're right, my bad. There are a lot of functions that are reformatted because they're indented to one more level. I hadn't noticed that.

About skipping the cache only when the source is a file, wouldn't it be simpler to do it this way than with the wrapper types? Something like e.g.

```diff
diff --git a/packages/sdk/src/definition/index.ts b/packages/sdk/src/definition/index.ts
index 83232bc..8843bde 100644
--- a/packages/sdk/src/definition/index.ts
+++ b/packages/sdk/src/definition/index.ts
@@ -603,11 +603,16 @@ export const introspect = {
                        return new MongoDBApi(schema, dataSources, fields, types, interpolateVariableDefinitionAsJSON);
                }),
        federation: introspectFederation,
-       openApi: async (introspection: OpenAPIIntrospection): Promise<RESTApi> =>
-               introspectWithCache(introspection, async (introspection: OpenAPIIntrospection): Promise<RESTApi> => {
+       openApi: async (introspection: OpenAPIIntrospection): Promise<RESTApi> => {
+               const generator = async (introspection: OpenAPIIntrospection): Promise<RESTApi> => {
                        const spec = loadOpenApi(introspection);
                        return await openApiSpecificationToRESTApiObject(spec, introspection);
-               }),
+               };
+               if (introspection.source.kind === 'file') {
+                       return generator(introspection);
+               }
+               return introspectWithCache(introspection, generator);
+       },
 };
 ```

**Comment by StarpTech** [comment]
> About skipping the cache only when the source is a file, wouldn't it be simpler to do it this way than with the wrapper types? Something like e.g.

Yes, my example should outline the approach to identifying different introspection types in a type-safe way.

**Pull request closed without merging**